### PR TITLE
Theme Discovery: Back to Discover navigation.

### PR DESCRIPTION
**Pull request by rcrdortiz**
Related to https://github.com/Automattic/wp-calypso/issues/82852

I've implemented a way to navigate back to the screen where we show the collection carousels when the user clicks on the `see all` button of any carousel.

This PR is only focused on the navigation functionality and not on the actual design of the `see all` screen. We can work on the design on a follow-up once we have a clear path forward. I did implement a basic UI that resembles the one found in our Figma designs.

Captura de pantalla 2023-10-16 a las 14 45 13

## Proposed Changes

* Created a new view for tiers. This view excludes some elements to resemble the Figma designs (see PR description)
* Changed the themes controller so that it's aware of the collection view.
* We can toggle between collection view and legacy view on tiers and subjects. This will work but we currently only have a title and description for the 2 existing carousels (well, the description isn't final).

## Testing Instructions

* Open the Calypso Live Link and open the LoTS with `?flags=themes/discovery-lots`
* Click on a `See all` button.
* You should see a new screen that displays the Collection.
* The screen should have a back button.
* When clicking on the back button you should be redirected to the theme showcase URL without any filters or subjects applied.

Before:
https://github.com/Automattic/wp-calypso/assets/1989914/1cbc3a9b-d12e-4667-b008-165cfd40c34c

After:
https://github.com/Automattic/wp-calypso/assets/1989914/469fcae2-80b2-4ee9-8afb-9e53f1f8c94e

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Review by dsas — COMMENTED** [review]
Will carry on reviewing later, but it all works well, the only thing I noticed so far is that the font changes between the two pages and I found that a little tiny bit jarring

**Comment by rcrdortiz** [comment]
> Will carry on reviewing later, but it all works well, the only thing I noticed so far is that the font changes between the two pages and I found that a little tiny bit jarring

Fonts are copied from the Figma designs, I could remove them for now until we implement the actual dark TS. Or I can update the fonts shown on the Collection Carousel screen.

File: client/my-sites/themes/theme-showcase.jsx [hunk 1]
```diff
@@ -337,6 +346,36 @@ class ThemeShowcase extends Component {
 		this.scrollToSearchInput();
 	};
 
+	getCollectionHeader = () => {
```
Comment (BogdanUngureanu) [inline client/my-sites/themes/theme-showcase.jsx:361]: This looks like a component to me. Maybe we should extract it from here in a separate file?

Comment (rcrdortiz) [reply client/my-sites/themes/theme-showcase.jsx:361]: We would need to prop drill a few things. In the future, it would be nice to extract most of the state in the theme showcase component to a React Context or Store and then create some components. At the time being it would just be a mess to create a component (we would need to pass props and functions from the ThemeShowcase to the new component). We also need to extract the URL creation logic to a separate component, so yeah. It would be nice to have it in a separate component but first, we would need to refactor the actual ThemeShowcase component, which is out of scope right now.

Comment (BogdanUngureanu) [reply client/my-sites/themes/theme-showcase.jsx:361]: We'll only need to pass `tier` (string) and `onClick` (callback) props; we don't have that many props to pass to the component.

Comment (rcrdortiz) [reply client/my-sites/themes/theme-showcase.jsx:361]: Since in the ThemeShowcase component, we already have some functions that return JSX, and the header is only used within the ThemeShowcase component (so it's not a reusable component) I see no need to extract it to a component (besides what I've already mentioned in my previous comment).

Comment (BogdanUngureanu) [reply client/my-sites/themes/theme-showcase.jsx:361]: I still think that extracting this in a separate component is better, but if you have strong feelings about it, let's leave it as it is.

> It would be nice to have it in a separate component but first, we would need to refactor the actual ThemeShowcase component

We don't need to do that; refactoring should be done incrementally. And passing two props wouldn't add much overhead. In the end, it's just two props.

> Since in the ThemeShowcase component, we already have some functions that return JSX, and the header is only used within the ThemeShowcase component (so it's not a reusable component) I see no need to extract it to a component (besides what I've already mentioned in my previous comment).

The component already has almost 600 LoC; personally, I find it hard to follow the code. Adding more code won't help with readability.

**Comment by BogdanUngureanu** [comment]
Screenshot 2023-10-16 at 17 58 13

Shouldn't the canonical be `https://wordpress.com/themes/premium`? 🤔

**Review by dsas — COMMENTED** [review]
> Fonts are copied from the Figma designs, I could remove them for now until we implement the actual dark TS. Or I can update the fonts shown on the Collection Carousel screen.

Whatever you think is best. I don't think it's a big deal either way, so long as they're in sync soon enough.

----

Another thing I noticed is that the `v=collection` doesn't follow all the way through. It's not a big blocking deal for me, but it is another paper-cut we need to follow up on. 

1. Go to /themes?flags=themes/discovery-lots
2. Click 'See all' on Premium Themes
3. Click on a theme
4. Click 'Back to themes'

I expected to go back to the collection i.e. `/themes/premium?v=collection` but instead the v= parameter is lost and I go back to `/themes/premium`.

----

**Comment by rcrdortiz** [comment]
> Another thing I noticed is that the `v=collection` doesn't follow all the way through. It's not a big blocking deal for me, but it is another paper-cut we need to follow up on.
> 
> 1. Go to /themes?flags=themes/discovery-lots
> 2. Click 'See all' on Premium Themes
> 3. Click on a theme
> 4. Click 'Back to themes'
> 
> I expected to go back to the collection i.e. `/themes/premium?v=collection` but instead the v= parameter is lost and I go back to `/themes/premium`.

True, I'll have a look.

**Comment by Copons** [comment]
I have an alternative proposal to MVP faster.

Since we are only preparing collections based on existing filters, how about on "See all" we open the _existing_ screen instead?

E.g.
- With `?flags=themes/discovery-lots`...
- From `/themes`, we click on the "See all" for the Premium Themes collection.
- It opens `/themes/premium`.

Building upon it with the feature flag, we could conditionally display the related title and collection, and point the "Back" button to the TS homepage (`/themes`).
Something like this:

Screenshot 2023-10-16 at 19 16 06

**Comment by rcrdortiz** [comment]
> I have an alternative proposal to MVP faster.
> 
> Since we are only preparing collections based on existing filters, how about on "See all" we open the _existing_ screen instead?
> 
> E.g.
> 
> * With `?flags=themes/discovery-lots`...
> * From `/themes`, we click on the "See all" for the Premium Themes collection.
> * It opens `/themes/premium`.
> 
> Building upon it with the feature flag, we could conditionally display the related title and collection, and point the "Back" button to the TS homepage (`/themes`). Something like this:
> 
> Screenshot 2023-10-16 at 19 16 06

That was the route I took while developing, but viewing premium themes for example would show the view you are suggesting regardless of if we're navigating from the carusel or the filters which seems wrong/broken from a user perspective. But as long as we agree on this I'm fine with it.

And regarding the design you're proposing I kinda dislike it, the information seems duplicated. What we could do though is to just show what we're currently showing in production and add a back button to the themes container.

And all these changes would look nothing like the Figma designs too, so that's something we have to keep in mind as well.

**Comment by rcrdortiz** [comment]
> I have an alternative proposal to MVP faster.
> 
> Since we are only preparing collections based on existing filters, how about on "See all" we open the _existing_ screen instead?
> 
> E.g.
> 
> * With `?flags=themes/discovery-lots`...
> * From `/themes`, we click on the "See all" for the Premium Themes collection.
> * It opens `/themes/premium`.
> 
> Building upon it with the feature flag, we could conditionally display the related title and collection, and point the "Back" button to the TS homepage (`/themes`). Something like this:
> 
> Screenshot 2023-10-16 at 19 16 06

Implemented it as you suggested, feel free to continue reviewing it.

**Comment by rcrdortiz** [comment]
> Screenshot 2023-10-16 at 17 58 13
> 
> Shouldn't the canonical be `https://wordpress.com/themes/premium`? 🤔

With the new approach, this shouldn't be an issue now. But I'll explain what I found about canonical URLs while I've developed this. I tried using:

```
const link = {
		rel: 'canonical',
		href,
	};
context.store.dispatch( setDocumentHeadLink( link ) );
```

Before rendering any content, which in theory should've changed the canonical URL, but it doesn't work, the canonical URL stays the same regardless (maybe SSR issue 🤷‍♂️ ). So in the end I opted for doing what we are currently doing which is to set the OG metas with the canonical URL.

Since we don't have the view param anymore the canonical URL will be the same (since we're back to only having one view).

**Comment by rcrdortiz** [comment]
As agreed in our weekly meeting I've reverted back to having the collection view.

File: client/my-sites/themes/theme-showcase-header.jsx [hunk 2]
```diff
@@ -50,6 +57,21 @@ export default function ThemeShowcaseHeader( { canonicalUrl, filter, tier, verti
...
+	}
+
+	if ( metaOnly ) {
+		return (
+			<>
+				<DocumentHead title={ documentHeadTitle } meta={ metas } />
+			</>
+		);
+	}
+
```
Comment (Copons) [inline client/my-sites/themes/theme-showcase-header.jsx:70]: Not a fan of `metaOnly` TBH.
It's not self-descriptive, and also slightly incorrect: it doesn't _only_ render `meta`s, but the whole `head`.

I wonder if it would be enough to accept that this is a collection view mod:

```jsx
<ThemeShowcaseHeader isCollectionView={ isCollectionView } />

// theme-showcase-header.jsx
if ( isCollectionView ) {
  metas.push( { name: 'robots', content: 'noindex' } );
  return <DocumentHead { ... } />;
}
```

--actually, I guess a `noIndex` prop might be helpful for our SEO problems. 🤔 
How about...
```jsx
<ThemeShowcaseHeader isCollectionView={ isCollectionView } noIndex={ isCollectionView } />
```

File: client/my-sites/themes/theme-showcase-header.jsx [hunk 3]
```diff
@@ -50,6 +57,21 @@ export default function ThemeShowcaseHeader( { canonicalUrl, filter, tier, verti
...
+			name: 'robots',
+			content: 'noindex',
+		} );
+	}
+
+	if ( metaOnly ) {
+		return (
+			<>
+				<DocumentHead title={ documentHeadTitle } meta={ metas } />
+			</>
```
Comment (Copons) [inline client/my-sites/themes/theme-showcase-header.jsx]: I don't think we need the empty fragment here, we can just return `DocumentHead` directly.

File: client/my-sites/themes/theme-showcase.jsx [hunk 4]
```diff
@@ -337,6 +358,45 @@ class ThemeShowcase extends Component {
...
+		switch ( this.props.tier ) {
+			case 'premium':
+				title = THEME_COLLECTIONS.premium.title;
+				description = THEME_COLLECTIONS.premium.description;
+				break;
+			case 'marketplace':
+				title = THEME_COLLECTIONS.partner.title;
+				description = THEME_COLLECTIONS.partner.description;
+				break;
+		}
```
Comment (Copons) [inline client/my-sites/themes/theme-showcase.jsx:374]: We should be more consistent in using `partner` or `marketplace`: pick one and stick with it.

For example, if we updated the [collections definitions](https://github.com/Automattic/wp-calypso/blob/29941e2f073e6fc48ad29194331db7f1d18dc560/client/my-sites/themes/collections/collection-definitions.tsx#L18) to use `marketplace` instead of `partner`, we could simplify this switch as:
```jsx
const { title, description } = THEME_COLLECTIONS[ this.props.tier ];
```

File: client/my-sites/themes/theme-showcase.scss [hunk 5]
```diff
@@ -262,6 +262,39 @@
...
+		}
+
+		.collection-header__description {
+			font-family: "SF Pro Text", serif;
+			font-size: 1rem;
+			font-style: normal;
+			font-weight: 400;
+			line-height: 24px;
+			letter-spacing: -0.32px;
+		}
```
Comment (Copons) [inline client/my-sites/themes/theme-showcase.scss:297]: Adding `@import "@automattic/typography/styles/variables";` we could use font size variables:

```diff
- font-size: 1rem;
+ font-size: $font-body;

- font-size: 2.75rem;
+ font-size: rem(44px);
```

I also think that we can scale the title responsively in a follow-up.
So, let's consider this whole comment as a note for a future improvement. 🙂

**Review by Copons — COMMENTED** [review]
I think this works really well and can't wait to confirm the SSR is working as expected!

I've left a couple of minor notes that I think should be revisited before merging, but this is pretty much ready to go. 🙂

File: client/my-sites/themes/theme-showcase.jsx (hunk 4 above)
Comment (Copons) [reply client/my-sites/themes/theme-showcase.jsx:374]: Nice, I've seen that this will change in #83166!
Feel free to ignore this comment, or use it for 83166. 🙂

File: client/my-sites/themes/theme-showcase-header.jsx (hunk 2 above)
Comment (rcrdortiz) [reply client/my-sites/themes/theme-showcase-header.jsx:70]: Sure, I'll change the prop name, we're already performing a check to add the noindex meta based on the noindex prop, so no need to change anything there.

Comment (rcrdortiz) [reply client/my-sites/themes/theme-showcase-header.jsx:70]: Updated.

File: client/my-sites/themes/theme-showcase-header.jsx (hunk 3 above)
Comment (rcrdortiz) [reply client/my-sites/themes/theme-showcase-header.jsx]: Updated.

**Review by Copons — APPROVED** [review]
